### DISCUSSION TITLE
[Share] QuickCopy items

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-actions-tags#165] [Action Scripts]
### Description

See https://github.com/windingwind/zotero-actions-tags/issues/161

Replace the "QuickCopy items" functions from Zutilo for alternative export translators, available from a right-click on one or multiple items, but also as a user defined shortcut:

> QuickCopy items: Copy selected items to the clipboard

https://github.com/wshanks/Zutilo/blob/master/docs/COMMANDS.md

### Event

None

### Operation

Script

### Data

```javascript
/**
 * QuickCopy items
 * @author Zotero Community
 * @usage Copy selected items to the clipboard
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/165
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/165
 */
if (items?.length === 0 || item) {
	return;
}

const Zotero = require("Zotero");
const ZoteroPane = require("ZoteroPane");

const pref = 'export.quickCopy.setting';
const origSetting = Zotero.Prefs.get(pref);

/**
 * Replace this line with your own setting
 * Steps:
 * 1. Change Settings -> Export -> Quick Copy -> Item Format
 * 2. Open Settings -> Advanced -> Config Editor
 * 3. Search for `export.quickCopy.setting`.
 * 4. Paste the value and replace the string `bibliography/html=http://www.zotero.org/styles/acm-siggraph` below.
 * 5. Reset Settings -> Export -> Quick Copy -> Item Format (If necessary)
 */
const newSetting = "bibliography/html=http://www.zotero.org/styles/acm-siggraph";

Zotero.Prefs.set(pref, newSetting);
ZoteroPane.copySelectedItemsToClipboard(false);
Zotero.Prefs.set(pref, origSetting);
return `QuickCopy done`;
```

**Comment by mjthoraval** [comment]
Thank you very much!

**Comment by thinkingaloudallowed** [comment]
Hi - I'm running into a problem with this scribt:

when I try to use it with Zotero 7 I get the following error message 

Script error: style getCiteProc is not a function 

looking round Zotero and GitHub in relation to this I found that there was an issue about not leaving the 'item format' empty in the export settings for Quick copy. I made sure to select this, but after triggering the script I found that the  item format was set to 'loading …' - in other words it is empty after using the script. 

Any suggestions?

**Reply by kerim** [reply]
I had the same issue, but then I followed the instructions in the middle of the script and it went away. Be sure to read the part starting /** carefully!

It might help if these instructions were at the top of the script instead of the middle?

**Comment by dm5n** [comment]
@dm5n
dm5n
Jan 16, 2024

Trying to copy the citekey for Markdown. Using Better Bibtex v6.7.143 with Zotero 7b on macOS 14.2.1.

Following the Steps in the Script:

    Selected 'Better BibTeX Citation Key Quick Copy' in Settings>Export
    Obtained a value of export=[alpha.numeric string]
    Substituted into const newSetting = "bibliography/html=http://www.zotero.org/styles/acm-siggraph";
    New line reads const newSetting = "export=[alpha.numeric string]";

Then to test: R-click on the Zotero item > TriggerAction > Citekey (the name I assigned to the script action)

But it copy+pastes the citeky in the format '\cite{citekey}'. How to change to just paste the citekey?

**Reply by mjthoraval** [reply]
You need to select "Cite Keys" in the settings of the BBT plugin:
![image](https://github.com/windingwind/zotero-actions-tags/assets/36054098/4b1773cb-0583-44ec-97a5-6f9bc339a1c1)

See the section "Drag and drop/hotkey citations" in [this BBT webpage](https://retorque.re/zotero-better-bibtex/citing/).
My reply in the forums: https://forums.zotero.org/discussion/comment/453978/#Comment_453978

**Comment by moritzklenk** [comment]
Hi, thanks very much for this script. Sorry, if what follows is a stupid question, but I ran into this issue:

when I use the script with a modified Chicago style, I get the following code copied to my clipboard:

`<div class="csl-bib-body" style="line-height: 1.35; margin-left: 2em; text-indent:-2em;">   <div class="csl-entry">Theweleit, Klaus. <i>Buch der Könige, Bd. 2y: Recording angels’ mysteries</i>. Sonderaufl. für Zweitausenseins. Basel: Stroemfeld/Roter Stern, 1996.&nbsp; &nbsp; ISBN: 978-3-87877-306-1.</div>   <span class="Z3988" title="url_ver=Z39.88-2004&amp;ctx_ver=Z39.88-2004&amp;rfr_id=info%3Asid%2Fzotero.org%3A2&amp;rft_id=urn%3Aisbn%3A978-3-87877-306-1&amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Abook&amp;rft.genre=book&amp;rft.btitle=Buch%20der%20K%C3%B6nige%2C%20Bd.%202y%3A%20Recording%20angels'%20mysteries&amp;rft.place=Basel&amp;rft.publisher=Stroemfeld%2FRoter%20Stern&amp;rft.edition=Sonderaufl.%20f%C3%BCr%20Zweitausenseins&amp;rft.aufirst=Klaus&amp;rft.aulast=Theweleit&amp;rft.au=Klaus%20Theweleit&amp;rft.date=1996&amp;rft.tpages=834&amp;rft.isbn=978-3-87877-306-1&amp;rft.language=German"></span> </div>`

Only if I paste it into a document in a text editor, I get what I was hoping for: Theweleit, Klaus. _Buch der Könige, Bd. 2y: Recording angels’ mysteries_. Sonderaufl. für Zweitausenseins. Basel: Stroemfeld/Roter Stern, 1996.   ISBN: 978-3-87877-306-1.

On a side note: I use Alfred and Keyboard Maestro - if that helps.

**Reply by mjthoraval** [reply]
As I understand, the HTML in the clipboard contains the formatting of the citation (italic, bold, ...), as "text/html":
https://forums.zotero.org/discussion/comment/422792/#Comment_422792
This formatting is automatically interpreted by the text editor when possible, after doing the normal paste `Ctrl`+`V`.

The clipboard also contains the text only, which is used when you do a `Ctrl`+`Shift`+`V`, as "text/plain".

See for example [here](https://evercoder.github.io/clipboard-inspector/) the content of your clipboard.

**Reply by moritzklenk** [reply]
Unfortunately not... using the clipboard inspector link you posted I get these results:

**for text/html:**
`<meta charset="UTF-8"><div class="csl-bib-body" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none; line-height: 1.35; margin-left: 2em; text-indent: -2em;"><div class="csl-entry">Theweleit, Klaus.<span class="Apple-converted-space"> </span><i>Buch der Könige, Bd. 2y: Recording angels’ mysteries</i>. Sonderaufl. für Zweitausenseins. Basel: Stroemfeld/Roter Stern, 1996.    ISBN: 978-3-87877-306-1.</div><span class="Z3988" title="url_ver=Z39.88-2004&amp;ctx_ver=Z39.88-2004&amp;rfr_id=info%3Asid%2Fzotero.org%3A2&amp;rft_id=urn%3Aisbn%3A978-3-87877-306-1&amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Abook&amp;rft.genre=book&amp;rft.btitle=Buch%20der%20K%C3%B6nige%2C%20Bd.%202y%3A%20Recording%20angels'%20mysteries&amp;rft.place=Basel&amp;rft.publisher=Stroemfeld%2FRoter%20Stern&amp;rft.edition=Sonderaufl.%20f%C3%BCr%20Zweitausenseins&amp;rft.aufirst=Klaus&amp;rft.aulast=Theweleit&amp;rft.au=Klaus%20Theweleit&amp;rft.date=1996&amp;rft.tpages=834&amp;rft.isbn=978-3-87877-306-1&amp;rft.language=German"></span></div>`

**for text/plain:**
```
<div class="csl-bib-body" style="line-height: 1.35; margin-left: 2em; text-indent:-2em;">
  <div class="csl-entry">Theweleit, Klaus. <i>Buch der Könige, Bd. 2y: Recording angels’ mysteries</i>. Sonderaufl. für Zweitausenseins. Basel: Stroemfeld/Roter Stern, 1996.&nbsp; &nbsp; ISBN: 978-3-87877-306-1.</div>
  <span class="Z3988" title="url_ver=Z39.88-2004&amp;ctx_ver=Z39.88-2004&amp;rfr_id=info%3Asid%2Fzotero.org%3A2&amp;rft_id=urn%3Aisbn%3A978-3-87877-306-1&amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Abook&amp;rft.genre=book&amp;rft.btitle=Buch%20der%20K%C3%B6nige%2C%20Bd.%202y%3A%20Recording%20angels'%20mysteries&amp;rft.place=Basel&amp;rft.publisher=Stroemfeld%2FRoter%20Stern&amp;rft.edition=Sonderaufl.%20f%C3%BCr%20Zweitausenseins&amp;rft.aufirst=Klaus&amp;rft.aulast=Theweleit&amp;rft.au=Klaus%20Theweleit&amp;rft.date=1996&amp;rft.tpages=834&amp;rft.isbn=978-3-87877-306-1&amp;rft.language=German"></span>
</div>
```

**Further information:**
I didn't have the issue with the zutilo quick copy alt function.  And I do not have the same result with the normal zotero quick copy, whatever style I use for this: it formats correctly.
It is only with this script that I receive these results.

**Reply by mjthoraval** [reply]
![image](https://github.com/windingwind/zotero-actions-tags/assets/36054098/7d6054c1-840b-4ea3-8b34-894d2b7358b1)
Do you have this unticked?

**Reply by moritzklenk** [reply]
my quick copy format in Zotero is "Better BiBteX citation quick copy" so there is not html to uncheck.

<img width="634" alt="Screenshot 2024-02-22 at 16 34 29" src="https://github.com/windingwind/zotero-actions-tags/assets/28430779/0fb15869-259a-45cf-87d0-1f733776b589">


but as I wrote: 

even if I chose, for example, "Chicago Manual of Style 17th edition" in zotero quick copy-- which I use for the actions and tags quick copy script above, I receive a properly formated clipboard. It looks like this, if I hand it over to some other script:
Theweleit, Klaus. 1996. _Buch der Könige, Bd. 2y: Recording angels’ mysteries._ Sonderaufl. für Zweitausenseins. Basel: Stroemfeld/Roter Stern

and like this if I paste it as plain text with cmd + shift + v:
Theweleit, Klaus. 1996. Buch der Könige, Bd. 2y: Recording angels’ mysteries. Sonderaufl. für Zweitausenseins. Basel: Stroemfeld/Roter Stern

so everthing is fine. 

_Except_ if I use the actions and tags script from above.
Then I get either the first for formated text or the second for plain text, and bot are code, not the bibliogrphy.

**Reply by mjthoraval** [reply]
Ok, I get it.
Just remove the `/html` after `bibliography` in `const newSetting`.

You probably had that option activated when you copied the content of `export.quickCopy.setting`, or you did not remove it from the example. The script actually takes the html option from there instead of the current settings.

**Reply by windingwind** [reply]
Try this if it still does not work, where the HTML option is force disabled when generating citation (don't forget to update the settings string):

```js
/**
 * QuickCopy items
 * @author Zotero Community
 * @usage Copy selected items to the clipboard
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/165
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/165
 */
if (items?.length === 0 || item) {
	return;
}

const Zotero = require("Zotero");
const ZoteroPane = require("ZoteroPane");
const Zotero_File_Interface = require("Zotero_File_Interface");

/**
 * Replace this line with your own setting
 * Steps:
 * 1. Change Settings -> Export -> Quick Copy -> Item Format
 * 2. Open Settings -> Advanced -> Config Editor
 * 3. Search for `export.quickCopy.setting`.
 * 4. Paste the value and replace the string `bibliography/html=http://www.zotero.org/styles/acm-siggraph` below.
 * 5. Reset Settings -> Export -> Quick Copy -> Item Format (If necessary)
 */
const newSetting = "bibliography/html=http://www.zotero.org/styles/american-chemical-society";

let format = Zotero.QuickCopy.getFormatFromURL(newSetting);
format = Zotero.QuickCopy.unserializeSetting(format);

items = items.filter(item => item.isRegularItem());

let locale = format.locale ? format.locale : Zotero.Prefs.get('export.quickCopy.locale');
		
Zotero_File_Interface.copyItemsToClipboard(
	items, format.id, locale, false, false
);
return `QuickCopy done`;
```

**Reply by moritzklenk** [reply]
removing the /html and replacing it by = did the job! Thank you very much for your help!
the line now reads:

`const newSetting = "bibliography=http://www.zotero.org/styles/chicago-fullnote-bibliography"`

this works perfectly.

A very satisfying problem and an even more satisfying solution.

**Reply by mjthoraval** [reply]
@windingwind: Maybe adding this information to the example above would be useful.
Maybe like this (or in another way):
```
const newSetting = "bibliography=http://www.zotero.org/styles/acm-siggraph";
// If you want to choose the HTML option, just add "/html" as in the example below:
// const newSetting = "bibliography/html=http://www.zotero.org/styles/acm-siggraph";
```

I guess having the default without the HTML option is what most people would be using.

**Comment by blindstitch** [comment]
> ### Description
> 
> See #161
> 
> Replace the "QuickCopy items" functions from Zutilo for alternative export translators, available from a right-click on one or multiple items, but also as a user defined shortcut:
> 
> > QuickCopy items: Copy selected items to the clipboard
> 
> https://github.com/wshanks/Zutilo/blob/master/docs/COMMANDS.md
> ### Event
> 
> None
> ### Operation
> 
> Script
> ### Data
> 
> ```js
> /**
>  * QuickCopy items
>  * @author Zotero Community
>  * @usage Copy selected items to the clipboard
>  * @link https://github.com/windingwind/zotero-actions-tags/discussions/165
>  * @see https://github.com/windingwind/zotero-actions-tags/discussions/165
>  */
> if (items?.length === 0 || item) {
> 	return;
> }
> 
> const Zotero = require("Zotero");
> const ZoteroPane = require("ZoteroPane");
> 
> const pref = 'export.quickCopy.setting';
> const origSetting = Zotero.Prefs.get(pref);
> 
> /**
>  * Replace this line with your own setting
>  * Steps:
>  * 1. Change Settings -> Export -> Quick Copy -> Item Format
>  * 2. Open Settings -> Advanced -> Config Editor
>  * 3. Search for `export.quickCopy.setting`.
>  * 4. Paste the value and replace the string `bibliography/html=http://www.zotero.org/styles/acm-siggraph` below.
>  * 5. Reset Settings -> Export -> Quick Copy -> Item Format (If necessary)
>  */
> const newSetting = "bibliography/html=http://www.zotero.org/styles/acm-siggraph";
> 
> Zotero.Prefs.set(pref, newSetting);
> ZoteroPane.copySelectedItemsToClipboard(false);
> Zotero.Prefs.set(pref, origSetting);
> return `QuickCopy done`;
> ```
> 
> ### Anything else
> 
> _No response_

I think steps 3 and 4 are written in a confusing way. I copied the value of the config setting and replaced newSetting (which for me appeared as `export=a515a220-6fef-45ea-9842-8025dfebcc8f`). Could you please clarify if that is what you are describing in those steps?

**Reply by mjthoraval** [reply]
The line after the comments becomes:
```javascript
const newSetting = "export=a515a220-6fef-45ea-9842-8025dfebcc8f";
```

**Comment by neybaf** [comment]
Thanks a lot !  I got some difficulties to make it works as  I wanted, but managed to do it.

**Comment by meefen** [comment]
Thanks for sharing the code. I finally got it to work after some tweaking. 

Steps 3 and 4 of the instruction were a bit confusing for me. The name of the config parameter was: `extensions.zotero.export.quickCopy.setting`. I suggest the following rewrite:


3. Search for `extensions.zotero.export.quickCopy.setting` and copy the value that looks like `export=xxxx-yyy-zzz`.
4. In the code below, replace the string `bibliography/html=http://www.zotero.org/styles/acm-siggraph` with the copied value.

**Comment by geogrrl** [comment]
I agree, steps 3 and 4, to me anyway, don't make sense, and it was very frustrating to try to make sense of it.

My first issue was in the config editor, you put in `export.quickCopy.setting` but it is not in there by default. However,  `extensions.zotero.export.quickCopy.setting` IS in there and 40 minutes later, I realize this is what is actually meant.

I've used this (awesome) plugin for a while and tbh I never had to read anything. I just plugged and played the code and luckily never had a problem. After coming to this thread and realising I didn't read and perform a bunch of actions, um, oops 😳. Anyway, here is another rewritten rendition of the instructions:

*Edit with your own settings:
* *Steps:*
* *1. Go to Zotero --> Settings -> Export -> Quick Copy -> Item Format (Select the item format you want)*
* *2. Go to Zotero Settings -> Advanced -> Config Editor*
* *3. Search for `extensions.zotero.export.quickCopy.setting`. Copy the setting.*
* *4. Replace the string `bibliography/html=http://www.zotero.org/styles/acm-siggraph` below with your setting*
* *5. Reset Settings -> Export -> Quick Copy -> Item Format (If necessary)*
*/*
`const newSetting = "export=ca65189f-8815-4afe-8c8b-8c7c15f0edca";`  (_example of a setting for betterbibtex_)

**Comment by gcarlet** [comment]
Since a few days this is not working for me... might be a problem with Zotero 8 ?

**Comment by Djoop** [comment]
Thanks for the code, is there a way when copying a single item, to use the "citation" CSL style/variant instead of the default <bibliography>?

**Comment by constwill22** [comment]
Hello. How can I configure the settings so that the document format does not display HTML and DOI?
